llms/mlx_lm/utils.py

```diff
@@ -390,17 +390,41 @@ def speculative_generate_step(
         kv_bits=kv_bits,
     )
 
+    def _process_and_sample(tokens, logits):
+        if logits_processors:
+            for processor in logits_processors:
+                logits = processor(tokens, logits)
+
+        logprobs = logits - mx.logsumexp(logits, keepdims=True)
+        logprobs = logprobs.squeeze(0)
+        y = sampler(logprobs)
+        return y, logprobs
+
     def _step(model, cache, y, n_predict=1):
         with mx.stream(generation_stream):
             logits = model(y[None], cache=cache)
             logits = logits[:, -n_predict:, :]
 
             quantize_cache_fn(cache)
-
-            logprobs = logits - mx.logsumexp(logits, keepdims=True)
-            logprobs = logprobs.squeeze(0)
-            y = sampler(logprobs)
-            return y, logprobs
+            if logits_processors:
+                nonlocal prev_tokens
+                out_y, out_logprobs = [], []
+                if n_predict > 1:
+                    y = y[: -(n_predict - 1)]
+                for i in range(n_predict):
+                    prev_tokens = (
+                        mx.concat([prev_tokens, y]) if prev_tokens is not None else y
+                    )
+                    y, logprobs = _process_and_sample(
+                        prev_tokens, logits[:, i : i + 1, :]
+                    )
+                    out_y.append(y)
+                    out_logprobs.append(logprobs)
```

Comment:
The unrolling here is a bit unfortunate.. but the logits processors can be autoregressive themselves (e.g. the `repetition_penalty`.